scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5240 @@
+#     "strategy" and refer to the means of calculating the quantity as the
+#     "strategy"?
+
+# Originally, I planned to filter out invalid distribution parameters;
+# distribution implementation functions would always work with "compressed",
+# 1D arrays containing only valid distribution parameters. There are two
+# problems with this:
+# - This essentially requires copying all arrays, even if there is only a
+#   single invalid parameter combination. This is expensive. Then, to output
+#   the original size data to the user, we need to "decompress" the arrays
+#   and fill in the NaNs, so more copying. Unless we branch the code when
+#   there are no invalid data, these copies happen even in the normal case,
+#   where there are no invalid parameter combinations. We should not incur
+#   all this overhead in the normal case.
+# - For methods that accept arguments other than distribution parameters, the
+#   user will pass in arrays that are broadcastable with the original arrays,
+#   not the compressed arrays. This means that this same sort of invalid
+#   value detection needs to be repeated every time one of these methods is
+#   called.
+# The much simpler solution is to keep the data uncompressed but to replace
+# the invalid parameters and arguments with NaNs (and only if some are
+# invalid). With this approach, the copying happens only if/when it is
+# needed. Most functions involved in stats distribution calculations don't
+# mind NaNs; they just return NaN. The behavior "If x_i is NaN, the result
+# is NaN" is explicit in the array API. So this should be fine.
+#
+# Currently, I am still leaving the parameters and function arguments
+# in their broadcasted shapes rather than, say, raveling. The intent
+# is to avoid back and forth reshaping. If authors of distributions have
+# trouble dealing with N-D arrays, we can reconsider this.
+#
+# Another important decision is that the *private* methods must accept
+# the distribution parameters as inputs rather than relying on these
+# cached properties directly (although the public methods typically pass
+# the cached values to the private methods). This is because the elementwise
+# algorithms for quadrature, differentiation, root-finding, and minimization
+# prefer that the input functions are strictly elementwise in the sense
+# that the value output for a given input element does not depend on the
+# shape of the input or that element's location within the input array.
+# When the computation has converged for an element, it is removed from
+# the computation entirely. As a result, the shape of the arrays passed to
+# the function will almost never be broadcastable with the shape of the
+# cached parameter arrays.
+#
+# I've sprinkled in some optimizations for scalars and same-shape/type arrays
+# throughout. The biggest time sinks before were:
+# - broadcast_arrays
+# - result_dtype
+# - is_subdtype
+# It is much faster to check whether these are necessary than to do them.
```

Comment:
This is something that I've done quite a lot myself, or as [Donald Knuth once put it](https://web.archive.org/web/20130731202547/http://pplab.snu.ac.kr/courses/adv_pl05/papers/p261-knuth.pdf):

> Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.